apps/discovery_api/lib/discovery_api/schemas/visualizations.ex

```diff
@@ -42,11 +45,48 @@ defmodule DiscoveryApi.Schemas.Visualizations do
     {:ok, existing_visualization} = get_visualization_by_id(id)
 
     if user.id == existing_visualization.owner_id do
+      visualization_changes = add_used_datasets(visualization_changes, Map.get(visualization_changes, :query))
+
       existing_visualization
       |> Visualization.changeset_update(visualization_changes)
       |> Repo.update()
     else
       {:error, "User does not have permission to update this visualization."}
     end
   end
+
+  defp add_used_datasets(visualization, nil), do: visualization
+
+  defp add_used_datasets(visualization, query) do
+    DiscoveryApi.prestige_opts()
+    |> Prestige.new_session()
+    |> Prestige.query("EXPLAIN (TYPE IO, FORMAT JSON) #{query}")
+    |> handle_explain_response(visualization)
+  end
+
+  defp handle_explain_response({:ok, response}, visualization) do
```

Comment:
Are some of these operations already done, to some extent, by the stuff under the authorization? Also, does there need to be some sort of authorization here too, so we don't leak private datasets via this field?